### PR TITLE
Implement different dropout strategies

### PR DESCRIPTION
**Pull request by xcgoner**
*Issue #, if available:*
https://github.com/apache/incubator-mxnet/issues/18669

*Description of changes:*
1. Fix a potential bug of ZoneoutCell implemented in MXNet. The new version is called RNNZoneoutCell, added as an option in DeepAR.
2. Add VariationalDropoutCell (similar to DropoutCell, but different time steps use the same mask, imported from GluonNLP) as an option in DeepAR.
3. Implement VariationalZoneoutCell (similar to ZoneoutCell, but different time steps use the same mask), which is , added as an option in DeepAR.

We train DeepAR with default hyperparameters, with Alpha Suffix iteration averaging of the last 40% iterations, and fix dropout rate as 0.1, and show the var, var/mean, std, std/mean, MASE, sMAPE, and RMSE on the "electricity" test set.
We run each configuration for 10 times.
For var ... std/mean, the variances are taken across all 10 trials and then we take the average over batches and time steps.
We can see that VariationalZoneoutCell achieves both smaller variance and MASE.

| dropout type           |    var |   var/mean |     std |   std/mean |     MASE |    sMAPE |    RMSE |
|------------------------|--------|------------|---------|------------|----------|----------|---------|
| ZoneoutCell            | 118470 |    2.30395 | 40.4088 |   0.104243 | 0.733318 | 0.104854 | 1645.47 |
| RNNZoneoutCell         | 126749 |    2.30689 | 41.4305 |   0.102345 | 0.71558  | 0.103203 | 1605.83 |
| VariationalDropoutCell | 123238 |    4.92394 | 40.1289 |   0.149108 | 0.702252 | 0.10295  | 1507.34 |
| VariationalZoneoutCell | 121027 |    2.08204 | 41.2225 |   0.10124  | 0.714836 | 0.103188 | 1501.22 |

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by xcgoner** [comment]
@benidis @szhengac I still cannot add reviewers, so I manually mention you two here.

File: src/gluonts/model/deepar/_estimator.py [hunk 1]
```diff
@@ -154,6 +155,16 @@ def __init__(
...
         assert num_cells > 0, "The value of `num_cells` should be > 0"
+        supported_dropoutcell_types = [
+            "ZoneoutCell",
+            "RNNZoneoutCell",
+            "VariationalDropoutCell",
+            "VariationalZoneoutCell",
+        ]
+        assert dropoutcell_type in supported_dropoutcell_types, (
+            "`dropoutcell_type` should be one of %s"
+            % (str(supported_dropoutcell_types))
```
Comment (lostella) [inline src/gluonts/model/deepar/_estimator.py:166]: Better to use f-strings:

```suggestion
            f"`dropoutcell_type` should be one of {supported_dropoutcell_types}"
```

File: src/gluonts/model/deepar/_estimator.py [hunk 2]
```diff
@@ -130,6 +130,7 @@ def __init__(
         num_layers: int = 2,
         num_cells: int = 40,
         cell_type: str = "lstm",
+        dropoutcell_type: str = "ZoneoutCell",
```
Comment (lostella) [inline src/gluonts/model/deepar/_estimator.py:133]: Could you add a description for this in the docstring above?

File: src/gluonts/mx/block/dropout.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    zoneout_states : float, default 0.
+        The dropout rate for state inputs on the first state channel.
+        Won't apply dropout if it equals 0.
+    
+    """
+
+    @validated()
+    def __init__(
+        self,
+        base_cell,
```
Comment (lostella) [inline src/gluonts/mx/block/dropout.py:171]:
```suggestion
        base_cell: RecurrentCell,
```

File: src/gluonts/mx/block/dropout.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    zoneout_states : float, default 0.
+        The dropout rate for state inputs on the first state channel.
+        Won't apply dropout if it equals 0.
+    
+    """
+
+    @validated()
+    def __init__(
+        self,
+        base_cell: HybridRecurrentCell,
```
Comment (lostella) [inline src/gluonts/mx/block/dropout.py:57]: `HybridRecurrentCell` or `RecurrentCell`? (see docstring above)

File: src/gluonts/mx/block/dropout.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    The mask is initialized when stepping forward for the first time and will remain
+    the same until .reset() is called. Thus, if using the cell and stepping manually without calling
+    .unroll(), the .reset() should be called after each sequence.
+
+    Parameters
+    ----------
+    base_cell : RecurrentCell
+        The cell on which to perform variational dropout.
+    zoneout_outputs : float, default 0.
```
Comment (lostella) [inline src/gluonts/mx/block/dropout.py:46]: Minor: elsewhere we use the convention of not annotating the type in the docstring (for input arguments), but just have the name on one line, followed by indented description (including default) in the folling lines

File: src/gluonts/mx/block/dropout.py (hunk 4 above)
Comment (xcgoner) [reply src/gluonts/mx/block/dropout.py:57]: RecurrentCell is the superclass of HybridRecurrentCell, which is more general. Maybe RecurrentCell is better.

**Comment by xcgoner** [comment]
@lostella I've resolved your previous comments and move references to bib. 
Could you help to check whether there is any other issues?

**Review by lostella — APPROVED** [review]
Thanks, looks good to me!
